src/middlewared/middlewared/plugins/cluster_linux/ctdb_public.py

```diff
@@ -0,0 +1,238 @@
+from middlewared.schema import Dict, IPAddr, Int, Str
+from middlewared.service import (accepts, private, job,
+                                 CRUDService, ValidationErrors,
+                                 CallError, filterable)
+from middlewared.utils import filter_list
+from middlewared.plugins.cluster_linux.utils import CTDBConfig
+import pathlib
+
+
+SHARED_VOL = CTDBConfig.CTDB_LOCAL_MOUNT.value
+PUB_IP_FILE = CTDBConfig.GM_PUB_IP_FILE.value
+ETC_IP_FILE = CTDBConfig.ETC_PUB_IP_FILE.value
+PUB_LOCK = CTDBConfig.PUB_LOCK.value
+
+
+class CtdbIpService(CRUDService):
+
+    class Config:
+        namespace = 'ctdb.public.ips'
+
+    @filterable
+    def query(self, filters=None, options=None):
+
+        ips = []
+        # logic is as follows:
+        #   1. if ctdb daemon is started
+        #       ctdb just reads the `ETC_IP_FILE` and loads the
+        #       ips written there into the cluster. However,
+        #       if a public ip is added/removed, it doesn't
+        #       mean the ctdb cluster has been reloaded to
+        #       see the changes in the file. So return what
+        #       the daemon sees.
+        #   2. if the `SHARED_VOL` is mounted and the `ETC_IP_FILE` exists
+        #       and is a symlink and the symlink is pointed to the
+        #       `PUB_IP_FILE` then read it and return the contents
+        if self.middleware.call_sync('service.started', 'ctdb'):
+            ips = self.middleware.call_sync('ctdb.general.ips')
+            ips = list(map(lambda i: dict(i, id=i['public_ip']), ips))
+        else:
+            try:
+                mounted = pathlib.Path(SHARED_VOL).is_mount()
+            except Exception:
+                # can happen when mounted but glusterd service
+                # is stopped/crashed etc
+                mounted = False
+
+            if mounted:
+                pub_ip_file = pathlib.Path(PUB_IP_FILE)
+                etc_ip_file = pathlib.Path(ETC_IP_FILE)
+                if pub_ip_file.exists():
+                    if etc_ip_file.is_symlink() and etc_ip_file.resolve() == pub_ip_file:
+                        with open(PUB_IP_FILE) as f:
+                            lines = f.read().splitlines()
+                            for i in lines:
+                                ips.append({
+                                    'id': i.split('/')[0],
```

Comment:
`id` from `ctdb -j ip` is sorted based on mildly complex rules specified in ctdb/tools/ctdb.c and ctdb/protocol/protocol_util.c `ctdb_sock_addr_cmp()`. Not sure if consistency in id allocation is important here (probably not).

Reply:
public IPs are not strictly handled like the private IPs are in ctdb. However, the `pnn` output from `ctdb ip -j` still seems to mirror the position in which the entry is written in the file.